// applicationinsights/out/Bootstrap/StatusLogger.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.StatusLogger = void 0;
var os = __importStar(require("os"));
var path = __importStar(require("path"));
var FileWriter_1 = require("./FileWriter");
var Constants_1 = require("../Declarations/Constants");
var StatusLogger = /** @class */ (function () {
    function StatusLogger(_writer, instrumentationKey) {
        if (_writer === void 0) { _writer = console; }
        if (instrumentationKey === void 0) { instrumentationKey = "unknown"; }
        this._writer = _writer;
        StatusLogger.DEFAULT_STATUS.Ikey = instrumentationKey;
    }
    StatusLogger.prototype.logStatus = function (data, cb) {
        if (typeof cb === "function" && this._writer instanceof FileWriter_1.FileWriter) {
            this._writer.callback = cb;
        }
        this._writer.log(data);
    };
    StatusLogger.DEFAULT_FILE_PATH = path.join(FileWriter_1.homedir, "status");
    StatusLogger.DEFAULT_FILE_NAME = "status_" + os.hostname() + "_" + process.pid + ".json";
    StatusLogger.DEFAULT_STATUS = {
        AgentInitializedSuccessfully: false,
        SDKPresent: false,
        Ikey: "unknown",
        AppType: "node.js",
        SdkVersion: Constants_1.APPLICATION_INSIGHTS_SDK_VERSION,
        MachineName: os.hostname(),
        PID: String(process.pid)
    };
    return StatusLogger;
}());
exports.StatusLogger = StatusLogger;
